File: source/downloads/code/LeetCode/Python/Longest-Increasing-Subsequence.py
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # n^2: 
        # dp[i]: the longest increasing subsequence
        dp = [1]*len(nums)
        for i in range(len(nums)):
            for j in range(i):
                if nums[i] > nums[j] and dp[i] < dp[j]+1:
                    dp[i] = dp[j] + 1
        rt = 0
        for i in range(len(nums)):
            if rt < dp[i]:
                rt = dp[i]
        return rt

        # n(logn)
        # dp: is an array to record the longest increasing subsequence
        dp = []
        for i in range(len(nums)):
            start = 0
            end = len(dp)-1
            # binary search dp, find the place dp[mid] < nums[i]
            while start <= end:
                mid = (start+end)/2
                if dp[mid] >= nums[i]:
                    end = mid - 1
                else:
                    start = mid + 1
            if start >= len(dp):
                dp.append(nums[i])
            else:
                dp[start] = nums[i]
        return len(dp)

File: source/downloads/code/LeetCode/Java/Longest-Increasing-Subsequence.java
public class Solution {
    public int lengthOfLIS(int[] nums) {
        if(nums.length==0) return 0;
        int []dp = new int[nums.length];
        // dp的实际长度
        int len = 0;
        for(int num: nums){
            // 在dp中找num
            int i = Arrays.binarySearch(dp, 0, len, num);
            if(i<0) i = -(i+1);
            // 更新dp
            dp[i] = num;
            if(i==len) len++;
        }
        return len;
    }
}
